### PR TITLE
docs(rfc): add api access modifiers style-guide RFC for v9

### PR DESCRIPTION
**Pull request by Hotell**
This RFC introduces style guide/ common approach for our packages API access modifiers that will be enforced by tooling.

[Preview 📖 👀](https://github.com/microsoft/fluentui/pull/23577/files?short_path=28b8ebf#diff-28b8ebf5a974c1a0ac4f3597aeb0b777c1b637688a0952b283f01c4c29f7b5f4)

[Tooling implementation PR 🛠](https://github.com/microsoft/fluentui/pull/23369) 

### 📣 Call to action to all teams

To be able to not delay v9 stable release any further we need all teams to review their packages by the end of Wednesday 22nd of June and apply following:

- if you use `@internal` modifier already, please verify that it is not exposed from package public API. If that's the case please remove it - BREAKING CHANGE in RC
- mark your package APIs with `@internal` modifier where necessary taking into consideration former bullet point

**Why is this needed?** 

As/If we introduce this style-guide/workflow, with aforementioned tooling PR being merged and  packages migrated to that new behavior, our packages will trim away `@internal`. 

if we won't catch this now, we will introduce BREAKING CHANGES in stable release which is a NO GO.

### CODE REVIEW

File: rfcs/shared/build-system/package-public-api.md
Comment (ling1726) [inline rfcs/shared/build-system/package-public-api.md:34]:
```suggestion
- tooling will trim the declaration from `/unstable` API surface.
```

File: rfcs/shared/build-system/package-public-api.md
Comment (ling1726) [inline rfcs/shared/build-system/package-public-api.md:205]: I think `_unstable` still has a use - for APIs that we want to allow some of our higher value partners to be able to take advantage of knowing full well that they will need to bear the responsibility for maintaining that, otherwise we would say that our composition model is internal 😅.

File: rfcs/shared/build-system/package-public-api.md
Comment (ling1726) [inline rfcs/shared/build-system/package-public-api.md:58]: I'm not quite sure how useful the prerelease stage modifiers will be since it's harder for partners to be aware of these tags, not all IDEs will render tsdoc the same way. For the scenarios that the prerelease modifiers cover, IMO `/unstable` entrypoint + `_unstable` suffix would be more impactful

**Review by ling1726 — APPROVED** [review]
I'm not fully convinced about the prerelease modifiers, but they are fully usable withour existing practices.

What I love is the ability to completely strip dts for fully internal APIs

File: rfcs/shared/build-system/package-public-api.md
Comment (spmonahan) [inline rfcs/shared/build-system/package-public-api.md:205]: I think we should keep it. Currently our composition API (e.g., `useComponent_unstable`) uses this in otherwise stable components. I could also see us adding other unstable APIs to stable components (e.g., `someProp_unstable`).

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:177]: >  it seems that everything is implicitly public unless otherwise annotated.?

yeah, API without annotation is implicitly considered as public ->
image

We can with enforce it with tooling, so missing annotation will fail on CI/build

image


>  Do we need this annotation? When would I use it?

It might feel like unnecessary churn for some devs to add this annotation all the time, on the other hand:
- it will force to provide reasonable amount of documentation to public APIs which is what we should be doing anyways
- consistency of using modifiers

personally I have no strong opinions on this, besides having consistency

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:205]: using `_unstable` api internally is not an issue. issue is if those apis are exposed as public package surface, which is already happening in our react-components "sub" packages

image

this RFC proposes to expose this packages only from `/unstable` api of particular package, which would be enforced by tooling. This would add us following benefits:
- no churn in package apis 
- stylistic thing: no  `_unstable` churn in identifiers anymore, we can use `_` suffix maybe to simplify (although prefix is a common pattern react hooks are playing against that unfortunately )

File: rfcs/shared/build-system/package-public-api.md
Comment (layershifter) [inline rfcs/shared/build-system/package-public-api.md:115]: I don't see any difference between in `@alpha` & `@beta` except semantics, is correct? As we export them as under unstable is there a sense to use `@unstable` instead?

IMO it will reduce maintenance cost as with current proposal we should ensure that `@alpha` will be changed to `@beta` when a package will be promoted. And we don't have any automation for it...

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:205]: sounds good, I added more context to Seans question: https://github.com/microsoft/fluentui/pull/23577#discussion_r899904711

**Review by layershifter — APPROVED** [review]

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:115]: > I don't see any difference between in @alpha & @beta except semantics, is correct 

correct. I added more context to one of Lings comments
↓
image

> As we export them as under unstable is there a sense to use @unstable instead?

that's great idea ! we would need to get this implemented to api-extractor / or eventually introduce custom tooling

File: rfcs/shared/build-system/package-public-api.md
Comment (layershifter) [reply rfcs/shared/build-system/package-public-api.md:58]: > additional context: if needed we can allow @beta to be shipped from stable and only @alpha from unstable but I'd leave that to team decisions ( IMO we should not do that ).

As `@beta` packages could have breaking changes IMO we should not ship them under stable.

To check my understanding: we have following chain for packages `alpha` -> `beta` -> `rc`. If we follow the same for APIs in comments, we should have also `@rc` comment, right? 👾

File: rfcs/shared/build-system/package-public-api.md
Comment (layershifter) [reply rfcs/shared/build-system/package-public-api.md:177]: IMO we should enforce it via ESLint to be sure that all APIs are properly annotated.

Comment (spmonahan) [reply rfcs/shared/build-system/package-public-api.md:177]: Thanks for the explanation. That really helped.

As you said, it is an extra bit of typing/work but I like the fact that _everything_ would then always be explicit. We uncovered quite a few accidentally public APIs during bug bash so I think it's worth being strict about.

File: rfcs/shared/build-system/package-public-api.md
Comment (spmonahan) [reply rfcs/shared/build-system/package-public-api.md:205]: Let me make sure my understanding is correct:

1. `import { MenuProvider } from '@fluentui/react-menu'`
2. `import { useMenuContext } from '@fluentui/react-menu/unstable'` (currently this is exported as `useMenuContext_unstable` from the stable import path)
3. This would be mirrored in `@fluentui/react-components`

I still think it's nice to have `_unstable` in the name for identifiers like `useMenuContext_unstable`. It's just that much more explicit and it's easy enough to find and replace or codemod when `_unstable` is dropped. The extra explicitness feels nice to me because our consumers should avoid our unstable APIs if at all possible. This issue isn't a blocker for me though.

**Review by spmonahan — APPROVED** [review]

File: rfcs/shared/build-system/package-public-api.md
Comment (GeoffCoxMSFT) [inline rfcs/shared/build-system/package-public-api.md:29]: We could have @public be the modifier and have the tooling make all things @internal by default.  This would force us to be explicit in what is exposed.  If we forget the attribute, the method or type can't get published.

**Review by GeoffCoxMSFT — APPROVED** [review]

File: rfcs/shared/build-system/package-public-api.md
Comment (ling1726) [reply rfcs/shared/build-system/package-public-api.md:29]: I actually agree with this 👍 how easy it is to implement might be a differnt story

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:58]: > does this mean that @fluentui/new-component will only be usable from @fluentui/new-component/unstable while the component is in alpha and beta stages ?

yes that's the idea (from TS api POV _

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:205]: > Let me make sure my understanding is correct:

yes thats correct 🙌

> I still think it's nice to have _unstable in the name for identifiers like useMenuContext_unstable

yeah sounds good, at least internally it will be more explicit that we are using unstable on our own. but really having a `_` would to the "trick" as well -> `useMenuContext_`

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:29]: interesting idea indeed @GeoffCoxMSFT ! that would work only for stable API's though, which is not our use-case as we have various stages of API/packages that are being shipped to users.

> how easy it is to implement might be a differnt story

haha YES

**Review by bsunderhus — APPROVED** [review]

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [inline rfcs/shared/build-system/package-public-api.md:203]: Please check out the summary with Fufure work paragraph - based on our discussion if it makes sense to you and if we  can proceed.

@layershifter @spmonahan @ling1726 

cheers!

File: rfcs/shared/build-system/package-public-api.md
Comment (Hotell) [reply rfcs/shared/build-system/package-public-api.md:115]: https://github.com/microsoft/rushstack/issues/3558

**Review by spmonahan — APPROVED** [review]

File: rfcs/shared/build-system/package-public-api.md
Comment (layershifter) [inline rfcs/shared/build-system/package-public-api.md:210]:
```suggestion
A["@alpha,@beta"] -- access from path --> U["@fluentui/package-name/unstable"]
B["@public"] -- access from path --> S["@fluentui/package-name"]
```

File: rfcs/shared/build-system/package-public-api.md
Comment (TristanWatanabe) [inline rfcs/shared/build-system/package-public-api.md:9]: > 
```suggestion
We want to control which parts of API surface of our packages are being exposed to public.
```

File: rfcs/shared/build-system/package-public-api.md
Comment (TristanWatanabe) [inline rfcs/shared/build-system/package-public-api.md:13]:
```suggestion
During v9 pre-release phase we ran into various situation where we didn't want to propagate APIs as ready to go, thus we established `_unstable` identifiers naming suffix to mark those as unstable. To complement this approach we also introduced `*/unstable` deep imports for API's that are not STABLE. This works to some extent but fails short as it's a manual work that is not checked by automation. This led to scenarios that we unintentionally exposed INTERNAL apis to public which led to breaking changes which we want to avoid at all costs as we follow proper semver.
```

File: rfcs/shared/build-system/package-public-api.md
Comment (TristanWatanabe) [inline rfcs/shared/build-system/package-public-api.md:21]:
```suggestion
We want to provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.
```

File: rfcs/shared/build-system/package-public-api.md
Comment (TristanWatanabe) [inline rfcs/shared/build-system/package-public-api.md:121]:
```suggestion
- tooling will NOT TRIM the declaration from `/unstable` API surface.
```

**Review by TristanWatanabe — APPROVED** [review]

**Comment by spmonahan** [comment]
@Hotell In a recent discussion the possibility of an unstable API or feature for an otherwise stable component came up. 

As a hypothetical let's say we need to add a new feature to `SpinButton`. Today `SpinButton` is stable in v9. We are unsure of either the exact API or the exact behavior (or perhaps both!) of this new feature but we want to be able to ship it so our users can test out the feature and provide feedback. How can we mark this feature as "unstable" so it's clear to users that it should be used with caution but available on an otherwise stable component?

I think this should be essentially the same flow as moving an unstable component to stable, just applied to individual features of a component.

I'm bringing this up now because it may be something we need in the future but I definitely don't think it should block this PR in any way.

**Pull request merged**
